include/dali/core/exec/tasking/scheduler.h

```diff
@@ -0,0 +1,330 @@
+  TaskFuture AddTask(SharedTask task) {
+    if (task->state_ != TaskState::New)
+      throw std::logic_error("A task can be submitted only once.");
+    auto res = task->results_;
+    AddTaskImpl(task);
+    return {std::move(task), std::move(res)};
+  }
+
+  /** Notifies the scheduler that a Waitable's state has changed and tasks waiting for it
+   *  may become ready.
+   */
+  void DLL_PUBLIC Notify(Waitable *w);
+
+  /** Waits for a task to complete. */
+  void Wait(Task *task);
+
+  /** Waits for a task to complete. */
+  void Wait(const SharedTask &task) {
+    Wait(task.get());
+  }
+
+  /** Makes all Pop functions return with an error value. */
+  void Shutdown() {
+    std::lock_guard g(mtx_);
+    shutdown_requested_ = true;
+    task_ready_.notify_all();
+    task_done_.notify_all();
+  }
+
+  /** Checks whether a shutdown was requested. */
+  bool ShutdownRequested() const {
+    return shutdown_requested_;
+  }
+
+ private:
+  /** Moves the task to the ready queue if all of its preconditions can be acquired.
+   *
+   * This function atomically checks that all preconditions can be met and if so, acquires them.
+   * If the preconditions where met, the task is moved from the pending list to the ready queue.
+   */
+  bool DLL_PUBLIC AcquireAllAndMoveToReady(SharedTask &task) noexcept;
+
+  void AddTaskImpl(SharedTask task) {
+    assert(task->state_ == TaskState::New);
+    if (task->Ready()) {  // if the task has no preconditions...
+      {
+        // ...then we add it directly to the read_ queue.
+        std::lock_guard lock(mtx_);
+        task->state_ = TaskState::Ready;
+        ready_.push(task);
+      }
+      task_ready_.notify_one();
+    } else {
+      // Otherwise, the task is added to the pending list
+      bool ready = false;
+      {
+        std::lock_guard lock(mtx_);
+        task->state_ = TaskState::Pending;
+        for (auto &pre : task->preconditions_) {
+          bool added = pre->AddToWaiting(task);
+          (void)added;
+          assert(added);
+        }
+        pending_.PushFront(task);
+        // ...and we check whether its preconditions are, in fact, met.
+        ready = AcquireAllAndMoveToReady(task);
+      }
+      if (ready)
+        task_ready_.notify_one();
+    }
+  }
+
+  friend class Task;
+
+  std::mutex mtx_;
+  std::condition_variable task_ready_, task_done_;
+
+  detail::TaskList pending_;
+  std::priority_queue<SharedTask, std::vector<SharedTask>, TaskPriorityLess> ready_;
+  bool shutdown_requested_ = false;
+};
+
+inline void Waitable::Notify(Scheduler &sched) {
+  sched.Notify(this);
+}
+
+inline void Task::Run(Scheduler &sched) {
+  assert(state_ == TaskState::Running);
+  wrapped_(this);
+  results_.clear();
+  inputs_.clear();
+  wrapped_ = {};
+  MarkAsComplete();
+  Notify(sched);
+  for (auto &r : release_) {
+    r->Release(sched);
+  }
+  release_.clear();
+  state_ = TaskState::Complete;
+}
+
+inline void TaskFuture::Wait(Scheduler &sched) {
```

Comment:
Now I have second thoughts regarding the API. Why does it need sched? Do you envision that it can be called in the context of multiple sheds?
For me it would be more intuitive to just
```
auto future = ex.AddTask(task3);
future.Value<void>();
```

Reply:
Storing the scheduler pointer inside the future itself isn't safe.
```C++
TaskFuture fut;
{
    Executor ex;
    ex.Start();
    fut = ex.AddTask(Task::Create([]() { return 42; });
}
fut.Value<void>();  // this would error out if `fut` object attempted to do `ex.Wait(fut.task)`.
```
It should be possible, if a bit wasteful, to implement it by saving the pointer to the scheduler inside the task object. That's one extra pointer in each and every task. Perhaps the price we're willing to pay.

Reply:
Up to you. It just seems not very intuitive now.

Reply:
Changed.